### PR TITLE
Merge utility crates.

### LINKED ISSUE #2005: Merge utility crates.
**Issue summary**

Forest contains several utility crates, often with a single module each. Release management for this many crates is troublesome so we would like for the crates to be merged together.
- `forest_utils` should be moved to the `utils` folder and be placed in the `forest_utils::io` module.
- `forest_hash_utils` should be placed in `forest_utils::hash`.
- `forest_json_utils` should be placed in `forest_utils::json`.
- `forest_net_utils` should be placed in `forest_utils::net`.
- `forest_test_utils` should be placed in `forest_utils::test`.
- `forest_macros` should be placed in `forest_utils::macros`.

**Other information and links**

### PR DESCRIPTION
**Pull request by tyshko5**
**Summary of changes**
Changes introduced in this pull request:
- 

**Reference issue to close (if applicable)**

Closes  #2005

**Other information and links**

### CODE REVIEW

**Review by lemmih — CHANGES_REQUESTED** [review]
Could you move `forest_utils` into the `utils/` folder and make a note of why `forest_hash_utils` and `forest_test_utils` couldn't be included in the crate?

**Comment by tyshko5** [comment]
`forest_test_utils` cannot be merged as it creates a dependency cycle
`forest_hash_utils ` cannot be merged as `fvm_ipfs_hamt` depends on it

**Comment by lemmih** [comment]
> `forest_test_utils` cannot be merged as it creates a dependency cycle

What cycle?

> `forest_hash_utils` cannot be merged as `fvm_ipfs_hamt` depends on it

They're depending on `v0.1` which has already been published. Why is it a problem that `fvm_ipld_hamt` needs `forest_hash_utils`?

**Comment by tyshko5** [comment]
Recreation of that cycle will take some effort, but it takes 6 packages in a dependency cycle.

**Review by lemmih — CHANGES_REQUESTED** [review]

File: utils/forest_utils/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,33 @@
+[package]
+name    = "forest_utils"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[package.metadata.docs.rs]
+features = ["test_constructors"]
```
[the remark this answers is not in the document]
Comment (tyshko5) [reply utils/forest_utils/Cargo.toml]: Removed

File: vm/actor_interface/src/adt/map.rs [hunk 2]
```diff
@@ -63,7 +63,10 @@ where
     }
 
     /// Returns a reference to the value corresponding to the key.
-    pub fn get<Q: ?Sized>(&self, k: &Q) -> Result<Option<&V>, AnyhowError>
+    pub fn get<Q: ?Sized + fil_actors_runtime_v8::fvm_ipld_hamt::Hash>(
```
[the remark this answers is not in the document]
Comment (tyshko5) [reply vm/actor_interface/src/adt/map.rs]: Fixed

File: utils/forest_utils/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+libc                       = "0.2"
+log.workspace              = true
+pbr.workspace              = true
+pin-project-lite.workspace = true
+thiserror.workspace        = true
+tokio                      = { workspace = true, features = ["fs"] }
+tokio-util                 = { version = "0.7.0", features = ["compat", "io-util"] }
+toml                       = "0.5"
+url.workspace              = true
+
```
[the remark this answers is not in the document]
Comment (tyshko5) [reply utils/forest_utils/Cargo.toml:23]: There were 2 empty lines before, I've deleted one. That's why it's not visible

File: vm/actor_interface/src/adt/map.rs [hunk 4]
```diff
@@ -4,8 +4,9 @@
 use crate::ActorVersion;
 use anyhow::Error as AnyhowError;
 use cid::Cid;
-use forest_hash_utils::{BytesKey, Hash};
 use forest_ipld_blockstore::BlockStore;
+use fvm_ipld_hamt::BytesKey;
+use fvm_ipld_hamt::Hash;
```
Comment (lemmih) [inline vm/actor_interface/src/adt/map.rs:9]:
```suggestion
use fvm_ipld_hamt::{BytesKey, Hash};
```

File: vm/actor_interface/src/adt/map.rs [hunk 5]
```diff
@@ -63,7 +64,7 @@ where
     }
 
     /// Returns a reference to the value corresponding to the key.
-    pub fn get<Q: ?Sized>(&self, k: &Q) -> Result<Option<&V>, AnyhowError>
+    pub fn get<Q: ?Sized + Hash>(&self, k: &Q) -> Result<Option<&V>, AnyhowError>
```
Comment (lemmih) [inline vm/actor_interface/src/adt/map.rs:67]: Do we need `Hash` here? If yes, why?

```suggestion
    pub fn get<Q: ?Sized>(&self, k: &Q) -> Result<Option<&V>, AnyhowError>
```

File: vm/actor_interface/Cargo.toml [hunk 6]
```diff
@@ -20,13 +20,13 @@ anyhow.workspace                 = true
...
 forest_fil_types.workspace       = true
-forest_hash_utils.workspace      = true
 forest_ipld_blockstore.workspace = true
 forest_json.workspace            = true
-forest_json_utils.workspace      = true
+forest_utils.workspace           = true
 fvm.workspace                    = true
 fvm_ipld_bitfield.workspace      = true
 fvm_ipld_encoding.workspace      = true
+fvm_ipld_hamt                    = "0.5"
```
Comment (lemmih) [inline vm/actor_interface/Cargo.toml:29]: `fvm_ipld_hamt` is part of the workspace.

```suggestion
fvm_ipld_hamt.workspace = true
```

**Review by lemmih — CHANGES_REQUESTED** [review]

**Review by lemmih — APPROVED** [review]

**Review by LesnyRumcajs — APPROVED** [review]

**Pull request merged**